share/expand_event_list_from_field.py

```diff
@@ -0,0 +1,47 @@
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+
+from typing import Any, Callable, Iterator, Optional
+
+import ujson
+
+ExpandEventListFromFieldExpanderCallable = Callable[
+    [dict[str, Any], int, int, str, str], Iterator[tuple[Any, int, bool, bool]]
+]
+
+
+class ExpandEventListFromFieldHelper:
+    def __init__(self, integration_scope: str, field_to_expand_event_list_from: str):
+        self.integration_scope: str = integration_scope
+        self.field_to_expand_event_list_from: str = field_to_expand_event_list_from
+
+    def expand(
+        self,
+        log_event: bytes,
+        json_object: Optional[dict[str, Any]],
+        starting_offset: int,
+        ending_offset: int,
+        expander: ExpandEventListFromFieldExpanderCallable,
+    ) -> Iterator[tuple[bytes, int, bool]]:
+        if json_object is None:
+            yield log_event, starting_offset, True
+        else:
+            for expanded_event, expanded_starting_offset, is_last_event_expanded, event_was_expanded in expander(
+                json_object,
+                starting_offset,
+                ending_offset,
+                self.integration_scope,
+                self.field_to_expand_event_list_from,
+            ):
+                if event_was_expanded:
+                    # empty values once json dumped will have a len() greater than 0, this will prevent
```

Comment:
it is indeed a side effect, but the fact that we must yield a string of bytes is part of the contract of the components involved: https://github.com/elastic/elastic-serverless-forwarder/blob/main/storage/storage.py#L55-L58

there are 2 steps where we can emit the original content of an event:
1. when fetching the content of the input (see https://github.com/elastic/elastic-serverless-forwarder/tree/main/storage, especially the decorators)
2. when and if we expand the events received from above

it is only in the case we have a json object with a field containing a list and we emits inner values of the list, that some type "juggling" happens. type of the unmarshaled value from the list is known only here and the information is lost through the chain of the components.
I don't see an alternative to carry this information over and finally accessing it in the component that filters empty event in order to design out of the system.

but this would be leaking some kind of two different domains in my opinion.
this is where the filtering happens: https://github.com/elastic/elastic-serverless-forwarder/blob/main/shippers/composite.py#L67-L75, where we deal only with the final forwarded event (https://github.com/elastic/elastic-serverless-forwarder/blob/main/handlers/aws/event.py)

Reply:
this makes sense - thanks for the explanation